Sources/OpenCombine/Publishers/Publishers.FlatMap.swift

```diff
@@ -0,0 +1,402 @@
+            }
+            return false
+        }
+
+        if shouldProcessQueue {
+            processQueue()
+        }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       completedWith completion: Subscribers.Completion<Failure>) {
+        switch completion {
+        case .finished:
+            removeActiveSubscription(forChild: child)
+        case .failure:
+            downstream?.receive(completion: completion)
+            deactivate()
+        }
+    }
+
+    private func removeActiveSubscription(forChild child: ChildSubscriber) {
+        evaluateActiveSubscriptionsExecuting { childSubscribers.remove(child) }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       receivedValue value: Child.Output) -> Subscribers.Demand {
+        // Obtain the lock in order to determine what to do
+        let (valueWasBuffered, processTheQueue): (Bool, Bool) = lock.do {
+            // Consider pending values to send when assessing downstreamDemand
+            if downstreamDemand - valuesToSend.count == .none {
+                valuesToSend.append((value, child))
+                return (valueWasBuffered: true, processTheQueue: false)
+            } else {
+                valuesToSend.append((value, nil))
+                if queueIsBeingProcessed {
+                    return (valueWasBuffered: false, processTheQueue: false)
+                }
+                queueIsBeingProcessed = true
+                return (valueWasBuffered: false, processTheQueue: true)
+            }
+        }
+
+        let demandResult = valueWasBuffered || demandForChild() == .unlimited
+            ? Subscribers.Demand.none : Subscribers.Demand.max(1)
+
+        if processTheQueue {
+            processQueue()
+        }
+
+        return demandResult
+    }
+
+    private func demandForChild() -> Subscribers.Demand {
+        return self.downstreamDemand == .unlimited ? .unlimited : .max(1)
+    }
+
+    private enum QueueWorkStatus {
+        case noWork
+        case sendFinish
+        case sendValues(values: ArraySlice<WorkType>)
+    }
+
+    private func processQueue() {
+        assert(queueIsBeingProcessed)
+
+        // We loop processing the queue in case somebody put stuff on the queue while we
+        // were sending values with the lock unlocked.
+        while true {
+            let work: QueueWorkStatus = lock.do {
+                if downstreamDemand == .none || valuesToSend.isEmpty {
+                    if sendFinishedAfterDrainingQueue && valuesToSend.isEmpty {
+                        return .sendFinish
+                    } else {
+                        queueIsBeingProcessed = false
+                        return .noWork
+                    }
+                }
+
+                let countToSend = min(valuesToSend.count, downstreamDemand.max ?? .max)
+                let result = valuesToSend[0..<countToSend]
+                valuesToSend.removeFirst(countToSend)
```

Comment:
This operation is O(n). For now let's leave it as is, but in the future we may need to use a better data structure than array here.

Can you add a `TODO`?

Reply:
Sure. I have to tell you, though, that I'm kind of shocked at the inconsistency of how Apple has been dealing with stuff related to demand. I may be missing something, but their zip implementation, for example seems completely bogus. Zip is pretty much the canonical example of the need for backpressure. Imagine zipping two sources. One fires constantly, and the other never. By setting a demand of 1 to each child, zip could avoid having to buffer a potentially infinite number of values from the chatty upstream. But Apple didn't bother to do that. Super strange.